experimental/zerotrie/src/builder/konst/builder.rs

```diff
@@ -0,0 +1,296 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use super::super::branch_meta::BranchMeta;
+use super::super::bytestr::ByteStr;
+use super::store::const_for_each;
+use super::store::ConstArrayBuilder;
+use super::store::ConstLengthsStack;
+use super::store::ConstSlice;
+use crate::error::Error;
+use crate::varint;
+
+/// A low-level builder for ZeroTrieSimpleAscii. Works in const contexts.
+pub(crate) struct ZeroTrieBuilderConst<const N: usize> {
+    data: ConstArrayBuilder<N, u8>,
+}
+
+impl<const N: usize> ZeroTrieBuilderConst<N> {
+    #[cfg(feature = "litemap")]
+    pub fn as_bytes(&self) -> &[u8] {
+        self.data.as_const_slice().as_slice()
+    }
+
+    pub const fn take_or_panic(self) -> [u8; N] {
+        self.data.const_take_or_panic()
+    }
+
+    pub const fn new() -> Self {
+        Self {
+            data: ConstArrayBuilder::new_empty([0; N], N),
+        }
+    }
+
+    #[must_use]
+    const fn prepend_ascii(self, ascii: u8) -> (Self, usize) {
```

Comment:
issue: eddyb pointed out that moving large arrays in const mode is a bad idea.

do you think we can do this with `&mut self` instead (both here and in other places containing a large const array)? it _ought_ to work in const. Ideally we do not have to move the const array after constructing it.

If `&mut self` doesn't work, we could also potentially change these abstractions to contain an `&mut [T; N]`. Helps with the destructor problem, too!

Reply:
I'd love to, but please help me make it compile.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1e9ef0aa821984ac8e66e67217cbac5b

Reply:
ah, oh well

Reply:
Are you okay if we move forward with the const builder based on moving big arrays, since there does not seem to be an alternative, and migrate to mut references if/when it becomes available in Rust?

Reply:
we should note a followup to fix this once this stuff stabilizes

Reply:
Yes, I'm okay with it, though it might be worth checking if the compile time impact is pathological

Reply:
Let's revisit this if we have capacity problems plugging it in with real databake. We can always fall back on the raw bytes constructor which doesn't move big arrays.

Reply:
wfm, open to having a followup about this